File: solutions/c/robot-simulator/1/robot_simulator.c
#include "robot_simulator.h"

robot_status_t robot_create(robot_direction_t direction, int x, int y)
{
    robot_status_t    robot =
    {
        direction,
        {x, y}
    };
    return robot;
}

void robot_move(robot_status_t *robot, const char *commands)
{
    for (size_t i = 0; i < strlen(commands); i++)
    {
        switch (commands[i])
        {
            case 'A':
                switch (robot->direction)
                {
                    case 0:
                        robot->position.y++;
                        break;
                    case 1:
                        robot->position.x++;
                        break;
                    case 2:
                        robot->position.y--;
                        break;
                    case 3:
                        robot->position.x--;
                        break;
                    default:
                        break;
                }
                break;
            case 'R':
                if (robot->direction == 3)
                    robot->direction = 0;
                else
                    robot->direction++;
                break;
            case 'L':
                if (robot->direction == 0)
                    robot->direction = 3;
                else
                    robot->direction--;
                break;
        }
    }
}

File: solutions/fsharp/robot-simulator/1/RobotSimulator.fs
module RobotSimulator

type Direction = North | East | South | West
type Position = int * int
type Robot = 
  { Direction: Direction
    Position: Position }

let create direction position = { Direction = direction; Position = position }

let rec move instructions robot = 
    let mutable direction = robot.Direction
    let mutable position = robot.Position
    for i in instructions do
        match i with
        | 'L' -> 
            match direction with
            | North -> direction <- West
            | East -> direction <- North
            | South -> direction <- East
            | West -> direction <- South
        | 'R' -> 
            match direction with
            | North -> direction <- East
            | East -> direction <- South
            | South -> direction <- West
            | West -> direction <- North
        | _ ->
            match direction with
            | North -> position <- (fst position, snd position + 1)
            | East -> position <- (fst position + 1, snd position)
            | South -> position <- (fst position, snd position - 1)
            | West -> position <- (fst position - 1, snd position)
    create direction position